data/pools/transactionPool.go

```diff
@@ -555,3 +559,133 @@ func (pool *TransactionPool) recomputeBlockEvaluator(committedTxIds map[transact
+// take longer than deadline to finish.
+func (pool *TransactionPool) AssembleBlock(round basics.Round, deadline time.Time) (*ledger.ValidatedBlock, error) {
+	start := time.Now()
+
+	pool.pendingMu.RLock()
+	pending := pool.pendingTxGroups
+	pendingCount := pool.pendingCountNoLock()
+	pool.pendingMu.RUnlock()
+
+	prev, err := pool.ledger.BlockHdr(round - 1)
+	if err != nil {
+		return nil, fmt.Errorf("could not make proposals at round %d: could not read block from ledger: %v", round, err)
+	}
+	newEmptyBlk := bookkeeping.MakeBlock(prev)
+
+	// Start the block evaluator, hinting its payset length based on the
+	// transaction pool's block evaluator
+	eval, err := pool.ledger.StartEvaluator(newEmptyBlk.BlockHeader, pendingCount)
+	if err != nil {
+		return nil, fmt.Errorf("could not make proposals at round %d: could not start evaluator: %v", round, err)
+	}
+
+	var stats telemetryspec.AssembleBlockMetrics
+	stats.StartCount = len(pending)
+	stats.StopReason = telemetryspec.AssembleBlockEmpty
+
+	// retrieve a list of all the previously known txid in the current round. We want to retrieve it here so we could avoid
+	// exercising the ledger read lock.
+	prevRoundTxIds := pool.ledger.GetRoundTxIds(round - 1)
+
+	for len(pending) > 0 {
+		txgroup := pending[0]
+		pending = pending[1:]
+
+		if len(txgroup) == 0 {
+			stats.InvalidCount++
+			continue
+		}
+
+		// if we already had this tx in the previous round, and haven't removed it yet from the txpool, that's fine.
+		// just skip that one.
+		if prevRoundTxIds[txgroup[0].ID()] {
+			stats.EarlyCommittedCount++
+			continue
+		}
+
+		if time.Now().After(deadline) {
+			stats.StopReason = telemetryspec.AssembleBlockTimeout
+			break
+		}
+
+		txgroupad := make([]transactions.SignedTxnWithAD, len(txgroup))
+		for i, tx := range txgroup {
+			txgroupad[i].SignedTxn = tx
+		}
+		err := eval.TransactionGroup(txgroupad)
+		if err == ledger.ErrNoSpace {
+			stats.StopReason = telemetryspec.AssembleBlockFull
+			break
+		}
+		if err != nil {
+			// GOAL2-255: Don't warn for common case of txn already being in ledger
+			switch err.(type) {
+			case ledger.TransactionInLedgerError:
+				stats.CommittedCount++
+			case transactions.MinFeeError:
+				stats.InvalidCount++
+				logging.Base().Infof("Cannot add pending transaction to block: %v", err)
+			default:
+				stats.InvalidCount++
+				logging.Base().Warnf("Cannot add pending transaction to block: %v", err)
+			}
+		}
+	}
+
+	lvb, err := eval.GenerateBlock()
+	if err != nil {
+		return nil, fmt.Errorf("could not make proposals at round %d: could not finish evaluator: %v", round, err)
+	}
+
+	// Measure time here because we want to know how close to deadline we are
+	dt := time.Now().Sub(start)
+	stats.Nanoseconds = dt.Nanoseconds()
+
+	if pool.logAssembleStats {
+		payset := lvb.Block().Payset
+		if len(payset) != 0 {
+			totalFees := uint64(0)
+
+			for i, txib := range payset {
+				fee := txib.Txn.Fee.Raw
+				encodedLen := txib.GetEncodedLength()
```

Comment:
I kinda bet that this is the slowest part of turning on logAssembleStats; maybe a TODO comment or a backlog issue for making a stats-lite mode that records the cheap stats?